src/main/scala/org/graphframes/lib/ConnectedComponents.scala

```diff
@@ -17,38 +17,262 @@
+    // ee.select(explode(
+    //    array(struct(col(SRC), col(DST)), struct(col(DST).as(SRC), col(SRC).as(DST)))).as(EDGE))
+    //  .select(col(s"$EDGE.$SRC").as(SRC), col(s"$EDGE.$DST").as(DST))
+
+    ee.unionAll(ee.select(col(DST).as(SRC), col(SRC).as(DST)))
+  }
+
+  /**
+   * Prepares the input graph for computing connected components by:
+   *   - de-duplicating vertices and assigning unique long IDs to each,
+   *   - changing edge directions to have increasing long IDs from src to dst,
+   *   - de-duplicating edges and removing loops.
+   * In the returned GraphFrame, the vertex DataFrame has two columns:
+   *   - column `id` stores a long ID assigned to the vertex,
+   *   - column `orig_id` stores the original ID.
+   * The edge DataFrame has two columns:
+   *   - column `src` stores the long ID of the source vertex,
+   *   - column `dst` stores the long ID of the destination vertex,
+   * where we always have `src` < `dst`.
+   */
+  private def prepare(graph: GraphFrame): GraphFrame = {
+    // TODO: This assignment job might fail if the graph is skewed.
+    val vertices = graph.indexedVertices
+      .select(col(LONG_ID).as(ID), col(ATTR))
+      // TODO: confirm the contract for a graph and decide whether we need distinct here
+      // .distinct()
+    val edges = graph.indexedEdges
+      .select(col(LONG_SRC).as(SRC), col(LONG_DST).as(DST))
+    val orderedEdges = edges.filter(col(SRC) !== col(DST))
+      .select(minValue(col(SRC), col(DST)).as(SRC), maxValue(col(SRC), col(DST)).as(DST))
+      .distinct()
+    GraphFrame(vertices, orderedEdges)
+  }
 
+  /**
+   * Returns the min neighbors of each vertex in a DataFrame with three columns:
+   *   - `src`, the ID of the vertex
+   *   - `min_nbr`, the min ID of its neighbors
+   *   - `cnt`, the total number of neighbors
+   *
+   * @return a DataFrame with three columns
+   */
+  private def minNbrs(ee: DataFrame): DataFrame = {
+    symmetrize(ee).groupBy(SRC).agg(min(col(DST)).as(MIN_NBR), count("*").as(CNT))
+  }
+
+  private def minValue(x: Column, y: Column): Column = {
+    when(x < y, x).otherwise(y)
+  }
+
+  private def maxValue(x: Column, y: Column): Column = {
+    when(x > y, x).otherwise(y)
+  }
+
+  /**
+   * Performs a possibly skewed join between edges and current component assignments.
+   * The skew join is done by broadcast join for frequent keys and normal join for the rest.
+   */
+  private def skewedJoin(
+      edges: DataFrame,
+      minNbrs: DataFrame,
+      broadcastThreshold: Int,
+      logPrefix: String): DataFrame = {
+    import edges.sqlContext.implicits._
+    val hubs = minNbrs.filter(col(CNT) > broadcastThreshold)
+      .select(SRC, MIN_NBR)
+      .as[(Long, Long)]
+      .collect()
+      .toMap // TODO: use OpenHashMap
+    if (hubs.isEmpty) {
+      return edges.join(minNbrs, SRC)
+    } else {
+      logger.debug(s"$logPrefix Number of skewed keys: ${hubs.size}.")
+    }
+    val isHub = udf { id: Long =>
```

Comment:
(discussed offline, here for the record:)  We could leave the hubs in a DataFrame instead of collecting them as a map.  Then the broadcast join could be done using a broadcast hint.  To be done eventually once we believe it is more efficient.